eloquentjs/4_3_list.js

```diff
@@ -0,0 +1,215 @@
+/**
+ * Returns list from array
+ * @param {*} inputArray 
+ * @returns 
+ */
+function arrayToList(inputArray) { 
+    if (!Array.isArray(inputArray)) { 
+        throw "input must be array";
+    }
+    if (inputArray.length === 0) {
+        return {value: null, rest: null};
+    }
+
+    const maxIndex = inputArray.length - 1;
+
+    function innerArrayToList(currentIndex, list = null) {
+        if (currentIndex < 0) {
+            return list;
+        }
+
+        const arrayElement = inputArray[currentIndex];
+
+        return innerArrayToList(currentIndex - 1, {value: arrayElement, rest: list});
+    }
+
+    return innerArrayToList(maxIndex);
+}
+
+/**
+ * Returns list from array
+ * @param {*} inputArray 
+ * @returns 
+ */
+function arrayToListIterative(inputArray) {
+    if (!Array.isArray(inputArray)) { 
+        throw "input must be array";
+    }
+    if (inputArray.length === 0) {
+        return {value: null, rest: null};
+    }
+
+    let listAcc = null;
+
+    for (let i = inputArray.length - 1; i >= 0; i--) {
+        const element = inputArray[i];
+        listAcc = {
+            value: element,
+            rest: listAcc
+        };
+    }
+
+    return listAcc;
+}
+
+/**
+ * Returns array from list
+ * @param {*} inputList 
+ * @returns 
+ */
+function listToArray(inputList) {
+    if (!isList(inputList)) {
+        throw "input must be list data structure";
+    }
+
+    function innerListToArray(list, arrayAccumulator = []) {
+        if(list.value !== null) {
+            arrayAccumulator.push(list.value);
+        }
+
+        if (list.rest === null) {
```

Comment:
This function might read better if you could call it with null as the value for the input list and your base case at the top was `if (list === null)...`